File: ChoInHeum/Week_5/boj_2589.py
from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(start_x, start_y, grid, h, w):
    visited = [[-1] * w for _ in range(h)]
    queue = deque()
    queue.append((start_x, start_y))
    visited[start_x][start_y] = 0

    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < h and 0 <= ny < w and grid[nx][ny] == 'L' and visited[nx][ny] == -1:
                visited[nx][ny] = visited[x][y] + 1
                queue.append((nx, ny))
                
    return visited

def find_max_distance(grid, h, w):
    max_dist = 0
    for i in range(h):
        for j in range(w):
            if grid[i][j] == 'L':
                visited = bfs(i, j, grid, h, w)
                for row in visited:
                    max_dist = max(max_dist, max(row))
    return max_dist

h, w = map(int, input().split())
grid = [list(input().strip()) for _ in range(h)]

print(find_max_distance(grid, h, w))

File: ChaeJangWoo/week5/boj_2589.cpp
#include <iostream>
#include <queue>
#include <algorithm>
#include <cstring>
using namespace std;

int N, M;
char map[51][51];
bool visited[51][51];
int dx[] = {1, -1, 0, 0};
int dy[] = {0, 0, 1, -1};

int bfs(int startY, int startX)
{
    memset(visited, false, sizeof(visited));
    queue<pair<pair<int, int>, int>> q; // ((y, x), 거리)
    q.push({{startY, startX}, 0});
    visited[startY][startX] = true;

    int maxDist = 0;

    while (!q.empty())
    {
        int y = q.front().first.first;
        int x = q.front().first.second;
        int dist = q.front().second;
        q.pop();

        maxDist = max(maxDist, dist);

        for (int i = 0; i < 4; ++i)
        {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny >= 0 && ny < N && nx >= 0 && nx < M)
            {
                if (!visited[ny][nx] && map[ny][nx] == 'L')
                {
                    visited[ny][nx] = true;
                    q.push({{ny, nx}, dist + 1});
                }
            }
        }
    }

    return maxDist;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    for (int i = 0; i < N; ++i)
    {
        cin >> map[i];
    }

    int answer = 0;
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < M; ++j)
        {
            if (map[i][j] == 'L')
            {
                answer = max(answer, bfs(i, j));
            }
        }
    }

    cout << answer << '\n';
    return 0;
}
